### PR TITLE
Add UTs and int tests for thread_notifications_lackey_dll

### PR DESCRIPTION
**Pull request by dcristoloveanu**
Add UTs and int tests for thread_notifications_lackey_dll

### CODE REVIEW

File: thread_notifications_lackey_dll/devdoc/thread_notifications_lackey_dll_requirements.md [hunk 1]
```diff
@@ -63,28 +65,28 @@ BOOL WINAPI DllMain(HINSTANCE hinstDLL, DWORD fdwReason, LPVOID lpvReserved);
...
+**SRS_THREAD_NOTIFICATIONS_LACKEY_DLL_01_005: [** If `fdwReason` is `DLL_PROCESS_ATTACH`: **]**
 
-- `DllMain` shall call `logger_init` to initialize logging for the module.
+- **SRS_THREAD_NOTIFICATIONS_LACKEY_DLL_01_006: [** `DllMain` shall call `logger_init` to initialize logging for the module. **]**
 
-- If `logger_init` fails, `DllMain` shall return FALSE.
+- **SRS_THREAD_NOTIFICATIONS_LACKEY_DLL_01_008: [** If `logger_init` fails, `DllMain` shall return FALSE. **]**
 
-- Otherwise, `DllMain` shall initialize the callback maintained by the module wth `NULL` and return `TRUE`.
+- **SRS_THREAD_NOTIFICATIONS_LACKEY_DLL_01_007: [** Otherwise, `DllMain` shall initialize the callback maintained by the module wth `NULL` and return `TRUE`. **]**
```
Comment (anporumb) [inline thread_notifications_lackey_dll/devdoc/thread_notifications_lackey_dll_requirements.md:74]: >wth [](http://example.com/codeflow?start=128&length=3)

with

**Comment by anporumb** [comment]
> typedef void (*THREAD_NOTIFICATION_LACKEY_DLL_CALLBACK_FUNC)(THREAD_NOTIFICATIONS_LACKEY_DLL_REASON reason);

header has plural here.

---
Refers to: thread_notifications_lackey_dll/devdoc/thread_notifications_lackey_dll_requirements.md:28 in 64edaa3. [](commit_id = 64edaa31999910fb05568316124052451f5a6880, deletion_comment = False)

**Comment by anporumb** [comment]
> #include "windows.h"

should come even before macro_utils.

---
Refers to: thread_notifications_lackey_dll/inc/thread_notifications_lackey_dll/thread_notifications_lackey_dll.h:10 in 64edaa3. [](commit_id = 64edaa31999910fb05568316124052451f5a6880, deletion_comment = False)

File: thread_notifications_lackey_dll/src/thread_notifications_lackey_dll.c [hunk 2]
```diff
@@ -1,28 +1,119 @@
...
+#include "c_logging/logger.h"
+
+#include "c_pal/interlocked.h"
+#include "c_pal/ps_util.h"
+
 #include "windows.h"
 
 #include "thread_notifications_lackey_dll/thread_notifications_lackey_dll.h"
 
-int thread_notifications_lackey_dll_init_callback(THREAD_NOTIFICATION_LACKEY_DLL_CALLBACK_FUNC thread_notification_cb)
```
Comment (anporumb) [inline thread_notifications_lackey_dll/src/thread_notifications_lackey_dll.c:7]: >int [](http://example.com/codeflow?start=0&length=3)

around here it should have a MU_DEFINE_ENUM_STRINGS...

File: thread_notifications_lackey_dll/src/thread_notifications_lackey_dll.c [hunk 3]
```diff
@@ -1,28 +1,119 @@
 // Copyright (C) Microsoft Corporation. All rights reserved.
 
+#include "c_logging/logger.h"
+
+#include "c_pal/interlocked.h"
+#include "c_pal/ps_util.h"
+
 #include "windows.h"
```
Comment (anporumb) [inline thread_notifications_lackey_dll/src/thread_notifications_lackey_dll.c:8]: should come before any of our headers.

File: thread_notifications_lackey_dll/src/thread_notifications_lackey_dll.c [hunk 4]
```diff
@@ -1,28 +1,119 @@
...
 
 #include "thread_notifications_lackey_dll/thread_notifications_lackey_dll.h"
 
-int thread_notifications_lackey_dll_init_callback(THREAD_NOTIFICATION_LACKEY_DLL_CALLBACK_FUNC thread_notification_cb)
+static volatile_atomic THREAD_NOTIFICATIONS_LACKEY_DLL_CALLBACK_FUNC notifications_cb;
```
Comment (anporumb) [inline thread_notifications_lackey_dll/src/thread_notifications_lackey_dll.c:12]: >notifications_cb [](http://example.com/codeflow?start=69&length=16)

for clarity should be initialized to NULL.

File: thread_notifications_lackey_dll/src/thread_notifications_lackey_dll.c [hunk 5]
```diff
@@ -1,28 +1,119 @@
...
+        result = MU_FAILURE;
+    }
+    else
+    {
+        /* Codes_SRS_THREAD_NOTIFICATIONS_LACKEY_DLL_01_002: [ Otherwise, thread_notifications_lackey_dll_init_callback shall initialize the callback maintained by the module with thread_notification_cb, succeed and return 0. ]*/
+        THREAD_NOTIFICATIONS_LACKEY_DLL_CALLBACK_FUNC existing_callback = (THREAD_NOTIFICATIONS_LACKEY_DLL_CALLBACK_FUNC)interlocked_compare_exchange_pointer((void* volatile_atomic*) & notifications_cb, (void*)thread_notifications_cb, NULL);
+        if (existing_callback != NULL)
+        {
+            /* Codes_SRS_THREAD_NOTIFICATIONS_LACKEY_DLL_01_003: [ If the callback was already initialized, thread_notifications_lackey_dll_init_callback shall fail and return a non-zero value. ] */
+            LogError("Invalid arguments: callback already set to %p", existing_callback);
```
Comment (anporumb) [inline thread_notifications_lackey_dll/src/thread_notifications_lackey_dll.c:31]: >Invalid arguments [](http://example.com/codeflow?start=22&length=17)

Not a case of invalid arguments... more like a user error. The only argument valid here is the previous value of notifications_cb.

Comment (anporumb) [inline thread_notifications_lackey_dll/src/thread_notifications_lackey_dll.c:31]: >LogError [](http://example.com/codeflow?start=12&length=8)

would be nice to log the address of the offending callback too.

File: thread_notifications_lackey_dll/src/thread_notifications_lackey_dll.c [hunk 6]
```diff
@@ -1,28 +1,119 @@
...
+        }
+    }
+
+    return result;
 }
 
 void thread_notifications_lackey_dll_deinit_callback(void)
 {
+    /* Codes_SRS_THREAD_NOTIFICATIONS_LACKEY_DLL_01_004: [ thread_notifications_lackey_dll_deinit_callback shall set the callback maintained by the module to NULL. ]*/
+    interlocked_exchange_pointer((void* volatile_atomic*)&notifications_cb, NULL);
```
Comment (anporumb) [inline thread_notifications_lackey_dll/src/thread_notifications_lackey_dll.c:46]: >interlocked_exchange_pointer [](http://example.com/codeflow?start=4&length=28)

(void*)

Comment (anporumb) [inline thread_notifications_lackey_dll/src/thread_notifications_lackey_dll.c:46]: >interlocked_exchange_pointer [](http://example.com/codeflow?start=4&length=28)

these 2 sequences of code have the same results,even when only 1 has balanced init/de_init calls:

```c
init(pf1)
deinit()
init(pf2) //succeds
```

versus
```c
init(pf1)
init(pf1)
deinit()
init(pf2) //succeds too
```

**Comment by anporumb** [comment]
>     THREAD_NOTIFICATIONS_LACKEY_DLL_REASON_ATTACH, \

When reading this from outside, it is unclear whether this is because of PROCESS or a THREAD attach (in Windows world both exist)

---
Refers to: thread_notifications_lackey_dll/inc/thread_notifications_lackey_dll/thread_notifications_lackey_dll.h:13 in 64edaa3. [](commit_id = 64edaa31999910fb05568316124052451f5a6880, deletion_comment = False)

File: thread_notifications_lackey_dll/src/thread_notifications_lackey_dll.c [hunk 7]
```diff
@@ -1,28 +1,119 @@
...
+
+    /* Codes_SRS_THREAD_NOTIFICATIONS_LACKEY_DLL_01_011: [ If fdwReason is DLL_THREAD_ATTACH: ]*/
+    case DLL_THREAD_ATTACH:
+    {
+        /* Codes_SRS_THREAD_NOTIFICATIONS_LACKEY_DLL_01_013: [ If a callback was passed by the user through a thread_notifications_lackey_dll_init_callback call, the callback shall be called with THREAD_NOTIFICATIONS_LACKEY_DLL_REASON_ATTACH. ]*/
+        THREAD_NOTIFICATIONS_LACKEY_DLL_CALLBACK_FUNC existing_callback = (THREAD_NOTIFICATIONS_LACKEY_DLL_CALLBACK_FUNC)interlocked_compare_exchange_pointer((void* volatile_atomic*) & notifications_cb, NULL, NULL);
+        if (existing_callback != NULL)
+        {
+            existing_callback(THREAD_NOTIFICATIONS_LACKEY_DLL_REASON_ATTACH);
+        }
```
Comment (anporumb) [inline thread_notifications_lackey_dll/src/thread_notifications_lackey_dll.c:89]: >} [](http://example.com/codeflow?start=8&length=1)

The else branch should log a miraculously thread_attach without process_attach before.

File: thread_notifications_lackey_dll/src/thread_notifications_lackey_dll.c [hunk 8]
```diff
@@ -1,28 +1,119 @@
...
+        /* Codes_SRS_THREAD_NOTIFICATIONS_LACKEY_DLL_01_015: [ DllMain shall return TRUE. ]*/
+        result = TRUE;
+        break;
+    }
+
+    /* Codes_SRS_THREAD_NOTIFICATIONS_LACKEY_DLL_01_009: [ If fdwReason is DLL_PROCESS_DETACH: ]*/
+    case DLL_PROCESS_DETACH:
+        /* Codes_SRS_THREAD_NOTIFICATIONS_LACKEY_DLL_01_010: [DllMain shall call logger_deinit and return TRUE.]*/
+        logger_deinit();
```
Comment (anporumb) [inline thread_notifications_lackey_dll/src/thread_notifications_lackey_dll.c:113]: >logger_deinit [](http://example.com/codeflow?start=8&length=13)

So if a process loads 2 dlls... in some random order, how many times is logger_init and logger_deinit called and what happens in the follogin sequence:

dll1: load
dll2: load
dll2: unload 
dll1: unload

Between dll2 and dll1 unload, executing code does it have a logger?

I suspect yes, but just asking.

File: thread_notifications_lackey_dll/tests/thread_notifications_lackey_dll_int/thread_notifications_lackey_dll_int.c [hunk 9]
```diff
@@ -0,0 +1,157 @@
...
+
+TEST_FUNCTION(thread_notifications_lackey_dll_init_callback_works)
+{
+    // arrange
+    
+    // act
+    // assert
+    ASSERT_ARE_EQUAL(int, 0, thread_notifications_lackey_dll_init_callback(test_thread_notifications_callback_1));
+    
+    // assert
```
Comment (anporumb) [inline thread_notifications_lackey_dll/tests/thread_notifications_lackey_dll_int/thread_notifications_lackey_dll_int.c:52]: >assert [](http://example.com/codeflow?start=7&length=6)

you mean "clean"

File: thread_notifications_lackey_dll/tests/thread_notifications_lackey_dll_int/thread_notifications_lackey_dll_int.c [hunk 10]
```diff
@@ -0,0 +1,157 @@
...
+            detach_count++;
+            break;
+        }
+        }
+    }
+
+    ASSERT_ARE_EQUAL(uint32_t, detach_count, attach_count);
+
+    // cleanup
+    thread_notifications_lackey_dll_deinit_callback();
```
Comment (anporumb) [inline thread_notifications_lackey_dll/tests/thread_notifications_lackey_dll_int/thread_notifications_lackey_dll_int.c:154]: >thread_notifications_lackey_dll_deinit_callback [](http://example.com/codeflow?start=4&length=47)

frankly, a wild deinit_callback suddenly appears.

File: thread_notifications_lackey_dll/tests/thread_notifications_lackey_dll_int/thread_notifications_lackey_dll_int.c [hunk 11]
```diff
@@ -0,0 +1,157 @@
...
+                }
+            }
+
+            if (j == i)
+            {
+                ASSERT_FAIL("Could not find attach for TID: %" PRIu32 "", thread_notifications[i].tid);
+            }
+            else
+            {
+                thread_notifications[j].tid = 0;
```
Comment (anporumb) [inline thread_notifications_lackey_dll/tests/thread_notifications_lackey_dll_int/thread_notifications_lackey_dll_int.c:142]: >thread_notifications[j].tid = 0; [](http://example.com/codeflow?start=16&length=32)

why is this line necessary here? Only asking because after collecting the data for the run, I believe that collected data (thread_notifications) shouldn't be touched by examining code.

File: thread_notifications_lackey_dll/tests/thread_notifications_lackey_dll_ut/thread_notifications_lackey_dll_mocked.c [hunk 12]
```diff
@@ -0,0 +1,11 @@
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+#define logger_init mock_logger_init
+#define logger_deinit mock_logger_deinit
+
+// Since we mightbe running in an environemtn where a DllMain exists (like running under VS CppUnittest)
```
Comment (anporumb) [inline thread_notifications_lackey_dll/tests/thread_notifications_lackey_dll_ut/thread_notifications_lackey_dll_mocked.c:6]: >environemtn [](http://example.com/codeflow?start=34&length=11)

environment

File: thread_notifications_lackey_dll/tests/thread_notifications_lackey_dll_ut/thread_notifications_lackey_dll_ut.c [hunk 13]
```diff
@@ -0,0 +1,350 @@
...
+
+/* Tests_SRS_THREAD_NOTIFICATIONS_LACKEY_DLL_01_001: [ If thread_notification_cb is NULL, thread_notifications_lackey_set_callback shall fail and return a non-zero value. ]*/
+TEST_FUNCTION(thread_notifications_lackey_dll_init_callback_with_NULL_thread_notification_cb_fails)
+{
+    // arrange
+    
+    // act
+    int result = thread_notifications_lackey_dll_init_callback(NULL);
+    
+    // assert
```
Comment (anporumb) [inline thread_notifications_lackey_dll/tests/thread_notifications_lackey_dll_ut/thread_notifications_lackey_dll_ut.c:85]: >assert [](http://example.com/codeflow?start=7&length=6)

missing ASSERT_ARE_EQUAL(char_ptr, umock_c_get_expected_calls(), umock_c_get_actual_calls());

here and everywhere

File: thread_notifications_lackey_dll/tests/thread_notifications_lackey_dll_ut/thread_notifications_lackey_dll_ut.c [hunk 14]
```diff
@@ -0,0 +1,350 @@
...
+
+    // act
+    thread_notifications_lackey_dll_deinit_callback();
+
+    // assert
+    // no explicit assert, no crash expected
+}
+
+/* Tests_SRS_THREAD_NOTIFICATIONS_LACKEY_DLL_01_004: [ thread_notifications_lackey_dll_deinit_callback shall set the callback maintained by the module to NULL. ]*/
+TEST_FUNCTION(thread_notifications_lackey_dll_deinit_callback_when_not_initialized_returns)
```
Comment (anporumb) [inline thread_notifications_lackey_dll/tests/thread_notifications_lackey_dll_ut/thread_notifications_lackey_dll_ut.c:136]: >thread_notifications_lackey_dll_deinit_callback_when_not_initialized_returns [](http://example.com/codeflow?start=14&length=76)

should pretty much have its own unittests, because _init has been called already at this point by other tests.

File: thread_notifications_lackey_dll/tests/thread_notifications_lackey_dll_ut/thread_notifications_lackey_dll_ut.c [hunk 15]
```diff
@@ -0,0 +1,350 @@
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+#include <stdbool.h>
+#include <inttypes.h>
+#include <stdlib.h>
+
+#include "macro_utils/macro_utils.h"
+#include "testrunnerswitcher.h"
```
Comment (anporumb) [inline thread_notifications_lackey_dll/tests/thread_notifications_lackey_dll_ut/thread_notifications_lackey_dll_ut.c:8]: >testrunnerswitcher [](http://example.com/codeflow?start=10&length=18)

missing #include "windows.h" because of `BOOL` usage

File: thread_notifications_lackey_dll/tests/thread_notifications_lackey_dll_ut/thread_notifications_lackey_dll_ut.c [hunk 16]
```diff
@@ -0,0 +1,350 @@
...
+
+    // act
+    BOOL result = DllMain_under_test(test_module_handle, DLL_THREAD_ATTACH, NULL);
+
+    // assert
+    ASSERT_IS_TRUE(result);
+    ASSERT_ARE_EQUAL(char_ptr, umock_c_get_expected_calls(), umock_c_get_actual_calls());
+
+    // cleanup
+    thread_notifications_lackey_dll_deinit_callback();
```
Comment (anporumb) [inline thread_notifications_lackey_dll/tests/thread_notifications_lackey_dll_ut/thread_notifications_lackey_dll_ut.c:235]: >thread_notifications_lackey_dll_deinit_callback [](http://example.com/codeflow?start=4&length=47)

in the real world, would here always be a 

test_thread_notifications_callback_1(THREAD_NOTIFICATIONS_LACKEY_DLL_REASON_DEATTACH) 

because "how does Elvis leave the building without getting out of the elevator first?"

File: thread_notifications_lackey_dll/tests/thread_notifications_lackey_dll_ut/thread_notifications_lackey_dll_ut.c [hunk 17]
```diff
@@ -0,0 +1,350 @@
...
+
+/* Tests_SRS_THREAD_NOTIFICATIONS_LACKEY_DLL_01_017: [ If fdwReason is any other value, DllMain shall terminate the process. ]*/
+// We don't really know that 0xFF won't be used in the future, but atm it is not
+TEST_FUNCTION(thread_notifications_lackey_DllMain_with_0xFF_terminates_the_process)
+{
+    // arrange
+    STRICT_EXPECTED_CALL(ps_util_terminate_process());
+
+    // act
+    ASSERT_IS_FALSE(DllMain_under_test(test_module_handle, 0xFF, NULL));
```
Comment (anporumb) [inline thread_notifications_lackey_dll/tests/thread_notifications_lackey_dll_ut/thread_notifications_lackey_dll_ut.c:344]: >0xFF [](http://example.com/codeflow?start=59&length=4)

Might I dare here suggest using `MU_DIFFERENT` macro? No guessing :)

File: thread_notifications_lackey_dll/tests/thread_notifications_lackey_dll_int/thread_notifications_lackey_dll_int.c (hunk 10 above)
Comment (anporumb) [inline thread_notifications_lackey_dll/tests/thread_notifications_lackey_dll_int/thread_notifications_lackey_dll_int.c:154]: >thread_notifications_lackey_dll_deinit_callback [](http://example.com/codeflow?start=4&length=47)

for an INT test that claims to have something to do with DllAttach-y things, this is lacking any dlls actually being loaded in the process.

Please consider having real dlls that are LoadLibrary'd /Free'd in the process and run the int tests with real dlls.

**Review by anporumb — APPROVED** [review]
:shipit:

File: thread_notifications_lackey_dll/devdoc/thread_notifications_lackey_dll_requirements.md (hunk 1 above)
Comment (dcristoloveanu) [reply thread_notifications_lackey_dll/devdoc/thread_notifications_lackey_dll_requirements.md:74]: Fixed.

File: thread_notifications_lackey_dll/src/thread_notifications_lackey_dll.c (hunk 2 above)
Comment (dcristoloveanu) [reply thread_notifications_lackey_dll/src/thread_notifications_lackey_dll.c:7]: Added

File: thread_notifications_lackey_dll/src/thread_notifications_lackey_dll.c (hunk 3 above)
Comment (dcristoloveanu) [reply thread_notifications_lackey_dll/src/thread_notifications_lackey_dll.c:8]: Moved

File: thread_notifications_lackey_dll/src/thread_notifications_lackey_dll.c (hunk 4 above)
Comment (dcristoloveanu) [reply thread_notifications_lackey_dll/src/thread_notifications_lackey_dll.c:12]: Added

File: thread_notifications_lackey_dll/src/thread_notifications_lackey_dll.c (hunk 5 above)
Comment (dcristoloveanu) [reply thread_notifications_lackey_dll/src/thread_notifications_lackey_dll.c:31]: Correct, I had copy pasted the thing above, fixed now.

Comment (dcristoloveanu) [reply thread_notifications_lackey_dll/src/thread_notifications_lackey_dll.c:31]: Added

File: thread_notifications_lackey_dll/src/thread_notifications_lackey_dll.c (hunk 6 above)
Comment (dcristoloveanu) [reply thread_notifications_lackey_dll/src/thread_notifications_lackey_dll.c:46]: void-ized

Comment (dcristoloveanu) [reply thread_notifications_lackey_dll/src/thread_notifications_lackey_dll.c:46]: Actually in

init(pf1) => success
init(pf1) => fail
deinit()
init(pf2) //success

Once the callback is initialized we do not care even if it is the same callback to set it. It is user error as the module is aimed to be as simple as possible.

I thought of doing more complicated stuff, but frankly it is better to keep it simple and just flag user error by failing.

File: thread_notifications_lackey_dll/src/thread_notifications_lackey_dll.c (hunk 7 above)
Comment (dcristoloveanu) [reply thread_notifications_lackey_dll/src/thread_notifications_lackey_dll.c:89]: I'm going to avoid doing any fancy stuff regarding logging in this module. That's because who knows what locks we'd take. I'm going to try to keep this path as lock avoiding as possible.

File: thread_notifications_lackey_dll/src/thread_notifications_lackey_dll.c (hunk 8 above)
Comment (dcristoloveanu) [reply thread_notifications_lackey_dll/src/thread_notifications_lackey_dll.c:113]: It has a logger, we allow nested logger init calls - we have a counter for it. That being said, it is extremely funny that you ask. Each dll that has logger linked will get its own logger state.

File: thread_notifications_lackey_dll/tests/thread_notifications_lackey_dll_int/thread_notifications_lackey_dll_int.c (hunk 9 above)
Comment (dcristoloveanu) [reply thread_notifications_lackey_dll/tests/thread_notifications_lackey_dll_int/thread_notifications_lackey_dll_int.c:52]: Haha, yes, fixed.

File: thread_notifications_lackey_dll/tests/thread_notifications_lackey_dll_int/thread_notifications_lackey_dll_int.c (hunk 11 above)
Comment (dcristoloveanu) [reply thread_notifications_lackey_dll/tests/thread_notifications_lackey_dll_int/thread_notifications_lackey_dll_int.c:142]: It is necessary in case we'd have the same tid twice.
As in :
attach 42
detach 42
...
attach 42
detach 42

I don't think there is anything prohibiting that from happening.
And I wanted that each detach to find its pair or else fail the test.

Comment (dcristoloveanu) [reply thread_notifications_lackey_dll/tests/thread_notifications_lackey_dll_int/thread_notifications_lackey_dll_int.c:142]: And added a comment.

File: thread_notifications_lackey_dll/tests/thread_notifications_lackey_dll_ut/thread_notifications_lackey_dll_mocked.c (hunk 12 above)
Comment (dcristoloveanu) [reply thread_notifications_lackey_dll/tests/thread_notifications_lackey_dll_ut/thread_notifications_lackey_dll_mocked.c:6]: Fixed.

File: thread_notifications_lackey_dll/tests/thread_notifications_lackey_dll_ut/thread_notifications_lackey_dll_ut.c (hunk 13 above)
Comment (dcristoloveanu) [reply thread_notifications_lackey_dll/tests/thread_notifications_lackey_dll_ut/thread_notifications_lackey_dll_ut.c:85]: Added

File: thread_notifications_lackey_dll/tests/thread_notifications_lackey_dll_ut/thread_notifications_lackey_dll_ut.c (hunk 14 above)
Comment (dcristoloveanu) [reply thread_notifications_lackey_dll/tests/thread_notifications_lackey_dll_ut/thread_notifications_lackey_dll_ut.c:136]: Added an extra test suite

File: thread_notifications_lackey_dll/tests/thread_notifications_lackey_dll_ut/thread_notifications_lackey_dll_ut.c (hunk 15 above)
Comment (dcristoloveanu) [reply thread_notifications_lackey_dll/tests/thread_notifications_lackey_dll_ut/thread_notifications_lackey_dll_ut.c:8]: Added

File: thread_notifications_lackey_dll/tests/thread_notifications_lackey_dll_ut/thread_notifications_lackey_dll_ut.c (hunk 16 above)
Comment (dcristoloveanu) [reply thread_notifications_lackey_dll/tests/thread_notifications_lackey_dll_ut/thread_notifications_lackey_dll_ut.c:235]: True, it won't do much in this case.
But added it anyway.

File: thread_notifications_lackey_dll/tests/thread_notifications_lackey_dll_ut/thread_notifications_lackey_dll_ut.c (hunk 17 above)
Comment (dcristoloveanu) [reply thread_notifications_lackey_dll/tests/thread_notifications_lackey_dll_ut/thread_notifications_lackey_dll_ut.c:344]: Great suggestion, applied.

File: thread_notifications_lackey_dll/tests/thread_notifications_lackey_dll_int/thread_notifications_lackey_dll_int.c (hunk 10 above)
Comment (dcristoloveanu) [reply thread_notifications_lackey_dll/tests/thread_notifications_lackey_dll_int/thread_notifications_lackey_dll_int.c:154]: The test aims at creating threads and making sure that the Dll forwards the notifications of the lackey dll, we don;t quite care about what other DLLs do with their own notifications ...

**Comment by dcristoloveanu** [comment]
> typedef void (*THREAD_NOTIFICATION_LACKEY_DLL_CALLBACK_FUNC)(THREAD_NOTIFICATIONS_LACKEY_DLL_REASON reason);

Would be sweet to claim that I did that on purpose, but I did not :-(.

---
In reply to: [1773098486](https://github.com/Azure/clds/pull/196#issuecomment-1773098486) [](http://example.com/codeflow?ancestors=1773098486)

---
Refers to: thread_notifications_lackey_dll/devdoc/thread_notifications_lackey_dll_requirements.md:28 in 64edaa3. [](commit_id = 64edaa31999910fb05568316124052451f5a6880, deletion_comment = False)

**Comment by dcristoloveanu** [comment]
> typedef void (*THREAD_NOTIFICATION_LACKEY_DLL_CALLBACK_FUNC)(THREAD_NOTIFICATIONS_LACKEY_DLL_REASON reason);

Fixed.

---
In reply to: [1773691644](https://github.com/Azure/clds/pull/196#issuecomment-1773691644) [](http://example.com/codeflow?ancestors=1773691644,1773098486)

---
Refers to: thread_notifications_lackey_dll/devdoc/thread_notifications_lackey_dll_requirements.md:28 in 64edaa3. [](commit_id = 64edaa31999910fb05568316124052451f5a6880, deletion_comment = False)

**Comment by dcristoloveanu** [comment]
> #include "windows.h"

Fixed.

---
In reply to: [1773099340](https://github.com/Azure/clds/pull/196#issuecomment-1773099340) [](http://example.com/codeflow?ancestors=1773099340)

---
Refers to: thread_notifications_lackey_dll/inc/thread_notifications_lackey_dll/thread_notifications_lackey_dll.h:10 in 64edaa3. [](commit_id = 64edaa31999910fb05568316124052451f5a6880, deletion_comment = False)

**Comment by dcristoloveanu** [comment]
>     THREAD_NOTIFICATIONS_LACKEY_DLL_REASON_ATTACH, \

Renamed to have THREAD_ATTACH/THREAD_DETACH

---
In reply to: [1773114803](https://github.com/Azure/clds/pull/196#issuecomment-1773114803) [](http://example.com/codeflow?ancestors=1773114803)

---
Refers to: thread_notifications_lackey_dll/inc/thread_notifications_lackey_dll/thread_notifications_lackey_dll.h:13 in 64edaa3. [](commit_id = 64edaa31999910fb05568316124052451f5a6880, deletion_comment = False)

**Review by avranju — APPROVED** [review]

**Review by anporumb — APPROVED** [review]
:shipit:

File: thread_notifications_lackey_dll/tests/thread_notifications_lackey_dll_ut/thread_notifications_lackey_dll_ut.c [hunk 18]
```diff
@@ -0,0 +1,355 @@
...
+}
+
+/* Tests_SRS_THREAD_NOTIFICATIONS_LACKEY_DLL_01_017: [ If fdwReason is any other value, DllMain shall terminate the process. ]*/
+TEST_FUNCTION(thread_notifications_lackey_DllMain_with_an_unknown_value_terminates_the_process)
+{
+    // arrange
+    STRICT_EXPECTED_CALL(ps_util_terminate_process());
+
+    // act
+    ASSERT_IS_FALSE(DllMain_under_test(test_module_handle, MU_DIFFERENT(DLL_THREAD_ATTACH, DLL_THREAD_DETACH, DLL_PROCESS_ATTACH, DLL_PROCESS_DETACH), NULL));
```
Comment (anporumb) [inline thread_notifications_lackey_dll/tests/thread_notifications_lackey_dll_ut/thread_notifications_lackey_dll_ut.c:349]: >MU_DIFFERENT [](http://example.com/codeflow?start=59&length=12)

pass to MU_DIFFERENT `THREAD_NOTIFICATIONS_LACKEY_DLL_REASON_VALUES` :)

Comment (dcristoloveanu) [reply thread_notifications_lackey_dll/tests/thread_notifications_lackey_dll_ut/thread_notifications_lackey_dll_ut.c:349]: Uhm, that is the enum that comes out of the module. Here I want to pass something different than the input values from Windows-land. And having an extra VALUES-list #define only to be used here is probably overkill. Thus I shall leave it as is.
